crypto/BRCryptoWalletManager.c

```diff
@@ -0,0 +1,1265 @@
+                cwmGetNonceAsETH,
+
+                // Events - Announce changes to entities that normally impact the UI.
+                cwmWalletManagerEventAsETH,
+                cwmPeerEventAsETH,
+                cwmWalletEventAsETH,
+                cwmEventTokenAsETH,
+                //       BREthereumClientHandlerBlockEvent funcBlockEvent;
+                cwmTransactionEventAsETH
+
+            };
+
+            // TODO: Race Here - WalletEvent before cwm->u.eth is assigned.
+            cwm->u.eth = ewmCreate (cryptoNetworkAsETH(network),
+                                    cryptoAccountAsETH(account),
+                                    (BREthereumTimestamp) cryptoAccountGetTimestamp(account),
+                                    (BREthereumMode) mode,
+                                    client,
+                                    cwmPath);
+
+            break;
+        }
+
+        case BLOCK_CHAIN_TYPE_GEN: {
+            break;
+        }
+    }
+
+    // TODO: cwm->wallet IS NOT assigned here; it is assigned in WalletEvent.created handler
+
+    free (cwmPath);
+
+    //    listener.walletManagerEventCallback (listener.context, cwm);  // created
+    //    listener.walletEventCallback (listener.context, cwm, cwm->wallet);
+
+    return cwm;
+}
+
+static void
+cryptoWalletManagerRelease (BRCryptoWalletManager cwm) {
+    printf ("Wallet Manager: Release\n");
+    cryptoAccountGive (cwm->account);
+    cryptoNetworkGive (cwm->network);
+    if (NULL != cwm->wallet) cryptoWalletGive (cwm->wallet);
+
+    for (size_t index = 0; index < array_count(cwm->wallets); index++)
+        cryptoWalletGive (cwm->wallets[index]);
+    array_free (cwm->wallets);
+
+    free (cwm->path);
+    free (cwm);
+}
+
+extern BRCryptoNetwork
+cryptoWalletManagerGetNetwork (BRCryptoWalletManager cwm) {
+    return cryptoNetworkTake (cwm->network);
+}
+
+extern BRCryptoAccount
+cryptoWalletManagerGetAccount (BRCryptoWalletManager cwm) {
+    return cryptoAccountTake (cwm->account);
+}
+
+extern BRSyncMode
+cryptoWalletManagerGetMode (BRCryptoWalletManager cwm) {
+    return cwm->mode;
+}
+
+extern BRCryptoWalletManagerState
+cryptoWalletManagerGetState (BRCryptoWalletManager cwm) {
+    return cwm->state;
+}
+
+private_extern void
+cryptoWalletManagerSetState (BRCryptoWalletManager cwm,
+                             BRCryptoWalletManagerState state) {
+    cwm->state = state;
+}
+
+extern const char *
+cryptoWalletManagerGetPath (BRCryptoWalletManager cwm) {
+    return cwm->path;
+}
+
+extern BRCryptoWallet
+cryptoWalletManagerGetWallet (BRCryptoWalletManager cwm) {
+    return cryptoWalletTake (cwm->wallet);
+}
+
+extern size_t
+cryptoWalletManagerGetWalletsCount (BRCryptoWalletManager cwm) {
+    return array_count (cwm->wallets);
+}
+
+extern BRCryptoWallet
+cryptoWalletManagerGetWalletAtIndex (BRCryptoWalletManager cwm,
+                                     size_t index) {
+    return cryptoWalletTake (cwm->wallets[index]);
```

Comment:
Are wallets expected to be added and/or removed? What's the story here from a concurrency POV.

Reply:
CORE-306